sdk/rust/nym-sdk/src/mixnet/native_client.rs

```diff
@@ -198,6 +260,195 @@ pub struct MixnetClientSender {
     packet_type: Option<PacketType>,
 }
 
+impl AsyncRead for MixnetClient {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf,
+    ) -> Poll<tokio::io::Result<()>> {
+        let mut codec = ReconstructedMessageCodec {};
+
+        if self._read.pending() {
+            return self.read_buffer_to_slice(buf, cx);
+        }
+
+        let msg = match self.as_mut().poll_next(cx) {
+            Poll::Ready(Some(msg)) => msg,
+            Poll::Ready(None) => return Poll::Ready(Ok(())),
+            Poll::Pending => return Poll::Pending,
+        };
+
+        // let mut buffer = BytesMut::new();
+
+        codec.encode(msg, &mut self._read.buffer).unwrap();
+
+        // = buffer.to_vec();
+
+        self.read_buffer_to_slice(buf, cx)
+    }
+}
+
+impl AsyncWrite for MixnetClient {
```

Comment:
imo `AsyncWrite` shouldn't be implemented on `MixnetClient` directly because conceptually it doesn't make a lot of sense. you're really just sending full messages as opposed to bytes. there should be some intermediate structure, like

```rust
struct MixnetClientAsyncWrite{
  recipient: Recipient
}
```

so then you'd have something like `fn async_write(&self, recipient: Recipient) -> MixnetClientAsyncWrite` on the client where you could just write your bytes your heart's content without having to do any encoding

Reply:
now, for any "higher" level applications that might need `AsyncRead + AsyncWrite` maybe we could just make a wrapper like
```rust
struct Wrapper {
  inner: MixnetClient,
  // or whatever fields we need
  write_recipient: Recipient,
}
```
with maybe even `Deref` trait for easier use

Reply:
Thats a very good point, but I wanted to avoid adapters here, I'm gonna give it some more though, maybe with the new InputMessageCodec we can get to just sending bytes in, and they discretley get converted to messages...